source/extensions/filters/common/ext_authz/check_request_utils.cc

```diff
@@ -119,15 +134,28 @@ void CheckRequestUtils::setHttpRequest(
 
   // Fill in the headers.
   auto* mutable_headers = httpreq.mutable_headers();
-  headers.iterate([mutable_headers](const Envoy::Http::HeaderEntry& e) {
+
+  headers.iterate([request_header_matchers, mutable_headers](const Envoy::Http::HeaderEntry& e) {
     // Skip any client EnvoyAuthPartialBody header, which could interfere with internal use.
     if (e.key().getStringView() != Headers::get().EnvoyAuthPartialBody.get()) {
       const std::string key(e.key().getStringView());
-      if (mutable_headers->find(key) == mutable_headers->end()) {
-        (*mutable_headers)[key] = std::string(e.value().getStringView());
+
+      if (request_header_matchers != nullptr) {
+        if (request_header_matchers->matches(key)) {
+          if (mutable_headers->find(key) == mutable_headers->end()) {
+            (*mutable_headers)[key] = std::string(e.value().getStringView());
+          } else {
+            // Merge duplicate headers.
+            (*mutable_headers)[key].append(",").append(std::string(e.value().getStringView()));
+          }
+        }
       } else {
-        // Merge duplicate headers.
-        (*mutable_headers)[key].append(",").append(std::string(e.value().getStringView()));
+        if (mutable_headers->find(key) == mutable_headers->end()) {
+          (*mutable_headers)[key] = std::string(e.value().getStringView());
+        } else {
+          // Merge duplicate headers.
+          (*mutable_headers)[key].append(",").append(std::string(e.value().getStringView()));
+        }
```

Comment:
This block is a duplicate of the above. Please combine them; I think the condition can be `if (request_header_matchers == nullptr || request_header_matchers->matches(key))`